### PR TITLE
HPCC-19584 Add support for new field filters into JHTree

### PR DESCRIPTION
**Pull request by richardkchapman**
Refactor SegMonitorList to put it behind an interface, and provide an
alternative implementation using new field filters.

Remove unused code from SegMonitorList and SegMonitors.

Moved test code out of rtlnewkey.cpp as it was preventing indexer from working
in Eclipse.

Default createValueSets to 1 in compiler.

Clean up IIndexReadContext interface to remove methods that should not have
been there.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [x] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Roxie regression suite completed, with -fcreateValueSets=0 and 1

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review

File: system/jhtree/jhtree.cpp [hunk 1]
```diff
@@ -1954,14 +1902,227 @@ IKeyCursor * CKeyCursor::fixSortSegs(unsigned sortFieldOffset)
...
+        queryFilter(field).setHigh(keyBuffer, recInfo.getFixedOffset(field));
+        field++;
+    }
+}
+
+unsigned IndexRowFilter::lastRealSeg() const
+{
+    return lastReal;
+}
+unsigned IndexRowFilter::lastFullSeg() const
```
Comment (ghalliday) [inline system/jhtree/jhtree.cpp]: whitespace: normally have newlines between functions.

File: rtl/eclrtl/rtlnewkey.cpp [hunk 2]
```diff
@@ -892,6 +947,75 @@ class ValueSet : public CInterfaceOf<IValueSet>
...
+    }
+
+    virtual void endRange(void *buffer, size32_t offset, const RtlTypeInfo &parentType) const override
+    {
+        dbgassertex(type.isFixedSize() && parentType.isFixedSize());
+        byte *ptr = ((byte *) buffer) + offset;
+        unsigned nextRange;
+        bool res = findForwardMatchRange(ptr, nextRange);
+        assertex(res);
+        queryTransition(nextRange*2+1)->setHigh(buffer, offset, type);  // MORE Do I need to set high remainder if subtype shorter than type?
```
Comment (ghalliday) [inline rtl/eclrtl/rtlnewkey.cpp]: I think this comment is out of date

File: rtl/eclrtl/rtlnewkey.cpp [hunk 3]
```diff
@@ -1239,9 +1396,198 @@ MemoryBuffer & SetFieldFilter::serialize(MemoryBuffer & out) const
...
+}
+
+StringBuffer & SingleFieldFilter::serialize(StringBuffer & out) const
+{
+    out.append(field).append("=[");
+    size32_t len;
+    rtlDataAttr text;
+    type.getUtf8(len, text.refstr(), value);
+    size32_t size = rtlUtf8Size(len, text.getstr());
+    if (type.isNumeric())  // MORE - code elsewhere in this class is assuming unsigned and big-endian...
```
Comment (ghalliday) [inline rtl/eclrtl/rtlnewkey.cpp]: I think no longer true.

File: rtl/eclrtl/rtlnewkey.cpp [hunk 4]
```diff
@@ -1239,9 +1396,198 @@ MemoryBuffer & SetFieldFilter::serialize(MemoryBuffer & out) const
...
+        return out.append(size, text.getstr());
+    else
+        return out.appendf("'%*s'", size, text.getstr());
+}
+
+
+MemoryBuffer & SingleFieldFilter::serialize(MemoryBuffer & out) const
+{
+    out.appendPacked(field).append('=');
+    out.appendPacked(1);  // Signifying a single transition - which in turn means a SingleFiedlFilter not a set.
```
Comment (ghalliday) [inline rtl/eclrtl/rtlnewkey.cpp]: typo fiedl

**Review by ghalliday — COMMENTED** [review]
A few minor comments

File: system/jhtree/jhtree.ipp [hunk 5]
```diff
@@ -170,21 +170,20 @@ public:
...
     CKeyIndex &key;
-    const SegMonitorList *segs;
+    const IIndexFilterList *filter;
     char *keyBuffer = nullptr;
     Owned<CJHTreeNode> node;
     unsigned int nodeKey;
 
     bool eof=false;
-    bool matched=false;
+    bool matched=false; //MORE - this should probably be renamed. It's tracking state from one call of lookup to the next.
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply system/jhtree/jhtree.ipp]: Not really.

File: rtl/eclrtl/rtlnewkey.cpp (hunk 4 above)
Comment (richardkchapman) [reply rtl/eclrtl/rtlnewkey.cpp]: Fixd

File: rtl/eclrtl/rtlnewkey.cpp (hunk 3 above)
Comment (richardkchapman) [reply rtl/eclrtl/rtlnewkey.cpp]: Comment removed

File: rtl/eclrtl/rtlnewkey.cpp (hunk 2 above)
Comment (richardkchapman) [reply rtl/eclrtl/rtlnewkey.cpp]: Comment removed

File: system/jhtree/jhtree.cpp (hunk 1 above)
Comment (richardkchapman) [reply system/jhtree/jhtree.cpp]: added

File: system/jhtree/jhtree.cpp [hunk 6]
```diff
@@ -1954,14 +1902,227 @@ IKeyCursor * CKeyCursor::fixSortSegs(unsigned sortFieldOffset)
...
+        if (idx)
+            out.append(',');
+        out.appendf("%s=", recInfo.queryName(idx));
+        filter.describe(out);
+    }
+}
+
+bool IndexRowFilter::matchesBuffer(const void *buffer, unsigned lastSeg, unsigned &matchSeg) const
+{
+    bool matched = true;
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply system/jhtree/jhtree.cpp]: removed

File: system/jhtree/jhtree.cpp [hunk 7]
```diff
@@ -88,34 +88,65 @@ MODULE_EXIT()
...
+        if (idx)
+            out.append(',');
+        out.appendf("%s=", recInfo.queryName(idx));
+        filter.describe(out, *recInfo.queryType(idx));
+    }
+}
+
+bool SegMonitorList::matchesBuffer(const void *buffer, unsigned lastSeg, unsigned &matchSeg) const
+{
+    bool matched = true;
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply system/jhtree/jhtree.cpp]: removed

File: system/jhtree/jhtree.cpp [hunk 8]
```diff
@@ -88,34 +88,65 @@ MODULE_EXIT()
...
+    if (numFilterFields())
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply system/jhtree/jhtree.cpp]: Changed

File: system/jhtree/jhtree.ipp [hunk 9]
```diff
@@ -240,4 +239,43 @@ public:
     CPartialKeyCursor(const CKeyCursor &from, unsigned sortFieldOffset);
     ~CPartialKeyCursor();
 };
+
+// Specialization of a RowFilter allowing us to use them from JHTree. Allowed to assume that all keyed fields are fixed size (for now!)
+
+class IndexRowFilter : public RowFilter, public CInterfaceOf<IIndexFilterList>
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply system/jhtree/jhtree.ipp]: Not sure about that. I tried but the change didn't fall out easily or make the resulting code cleaner in my opinion. This class is - as the comment suggests - a specialization of RowFilter - in particular it makes assumptions about what is going on in the rowfilter - such as that the field numbers and the filter indexes correspond - which would be less clear if the RowFilter was a member.

**Pull request merged**